src/Modules/UngradedModules.jl

```diff
@@ -8712,3 +8716,258 @@ function dual(f::ModuleFPHom{<:ModuleFP, <:ModuleFP, Nothing}; # Third parameter
+
+  o = default_ordering(M)
+  LM = leading_module(Mq,o)
+
+  has_monomials_on_all_axes(LM) || error("not a finite dimensional vector space")
+  
+  d = 0
+  all_mons=[]
+  temp_mons = vector_space_basis(M,0)
+
+  while length(temp_mons) > 0
+    append!(all_mons,temp_mons)
+    d = d+1
+    temp_mons=vector_space_basis(M,d)
+  end
+
+  return all_mons
+end
+
+function vector_space_basis(M::SubquoModule,d::Int64)
+  R = base_ring(M)
+  F = ambient_free_module(M)
+  Mq,_ = sub(F,rels(M))
+
+  ambient_representatives_generators(M) == gens(F) || error("not implemented for M/N with non-trivial M")
+
+  o = default_ordering(M)
+  LM = leading_module(Mq,o)
+
+  return [x*e for x in all_monomials(R, d) for e in gens(F) if !(x*e in LM)]
+end
+
+function vector_space_basis(M::SubquoModule{T}
+  ) where {T<:MPolyLocRingElem{<:Field, <:FieldElem, <:MPolyRing, <:MPolyRingElem, 
+                               <:MPolyComplementOfKPointIdeal}}
+  F = ambient_free_module(M)
+  Mq,_ = sub(F,rels(M))
+
+  M_shift,_,_ = shifted_module(Mq)
+  if isdefined(F,:ordering) && is_local(F.ordering)
+    o = F.ordering
+  else
+    o = negdegrevlex(base_ring(M_shift))*lex(ambient_free_module(M_shift))
+  end
+  LM = leading_module(M_shift,o)
+
+  return vector_space_basis(quo(ambient_free_module(LM),gens(LM))[1])
+end
+
+function vector_space_basis(M::SubquoModule{T},d::Int64
+  ) where {T<:MPolyLocRingElem{<:Field, <:FieldElem, <:MPolyRing, <:MPolyRingElem, 
+                               <:MPolyComplementOfKPointIdeal}}
+  F = ambient_free_module(M)
+  Mq,_ = sub(F,rels(M))
+
+  M_shift,_,_ = shifted_module(Mq)
+  if isdefined(F,:ordering) && is_local(F.ordering)
+    o = F.ordering
+  else
+    o = negdegrevlex(base_ring(M_shift))*lex(ambient_free_module(M_shift))
+  end
+  LM = leading_module(M_shift,o)
+
+  return vector_space_basis(quo(ambient_free_module(LM),gens(LM))[1],d)
+end
+
+function vector_space_basis(M::SubquoModule{T}
+  ) where {T<:MPolyLocRingElem}
+  error("only available in global case and for localization at a point")
+end
+
+function vector_space_basis(M::SubquoModule{T},d::Int64
+  ) where {T<:MPolyLocRingElem}
+  error("only available in global case and for localization at a point")
+end
+
+@doc raw"""
+    has_monomials_on_all_axes(M::SubquoModule)
+
+Internal function to test whether M is finite-dimensional vector space. Do not use directly
+"""
+function has_monomials_on_all_axes(M::SubquoModule)
+  R = base_ring(M)
+
+  length(rels(M)) == 0 || error("not implemented for quotients")
+  
+  ambient_rank = ngens(ambient_free_module(M))
+  genlist = ambient_representatives_generators(M)
+  explist = Tuple{Vector{Int64}, Int64, Int}[]
+  for x in genlist
+    tempexp = leading_exponent(x)
+    tempdeg = sum(tempexp[1])
+    push!(explist,(tempexp[1],tempexp[2],tempdeg))
+  end
+  all(findfirst(x -> (x[1][i] == x[3] && x[2]==j),explist) != nothing for i in 1:ngens(R) for j in 1:ambient_rank) || return false
+  return true
```

Comment:
Could be rewritten as
```suggestion
  return all(findfirst(x -> (x[1][i] == x[3] && x[2]==j),explist) != nothing for i in 1:ngens(R) for j in 1:ambient_rank)
```
That said, I personally find it difficult to read these long constructs, and personally think that the explicit imperative version is easier to understand and debug -- but I'll not complain either way, use whatever you prefer.
```suggestion
  for i in 1:ngens(R), j in 1:ambient_rank
    if findfirst(x -> (x[1][i] == x[3] && x[2]==j), explist) === nothing
      return false
    end
  end
  return true
```
or for that matter
```suggestion
  for i in 1:ngens(R), j in 1:ambient_rank
    if !any(x -> (x[1][i] == x[3] && x[2]==j), explist)
      return false
    end
  end
  return true
```